project-koku/koku
# Cost-2888: Add exclude to our querysets

## Jira Ticket

[COST-2888](https://issues.redhat.com/browse/COST-2888)

## Description

Add the ability to do negative filtering to all of our endpoints.

## Testing
1. Turn Gunicorn on so that way the unleash client is initialized on the koku server with the `RUN_GUNICORN` environment variable. If you forget this step you will see something like this in the logs:
```
koku_server       | [2022-09-20 19:09:07,309] WARNING None Returning default value for feature: cost-enable-negative-filtering
koku_server       | [2022-09-20 19:09:07,310] WARNING None Attempted to get feature_flag cost-enable-negative-filtering, but client wasn't initialized!
```

If you have gunicorn problems like me you can add these vars as a work around to get ya moving: 
- https://github.com/project-koku/koku/pull/3890/commits/57b6f4d416f23fa0fa79493a1e8111e8913a300c
2. Upload dummy data.
```
make create-test-customer-data
make load-test-customer-data
```
3. Add `acct1234567` to the unleash flag here: http://localhost:4242/#/features
4. I have found the best way to test this is to choose a group by option with the monthly breakdown filters on an endpoint, example:
- http://localhost:8000/api/cost-management/v1/reports/gcp/costs/?group_by[gcp_project]=*&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly

I usually make a note of the overall cost & then pick out a specific option and note its cost as well:
```
overall: 5745347.77975477,
example-project-id:  5744047.59912087
```

5. Then add the exclude parameter to the url targeting the value you marked for exclusion:
http://localhost:8000/api/cost-management/v1/reports/gcp/costs/?group_by[gcp_project]=*&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&exclude[gcp_project]=example-project-id

**Check the Logs:**
```
koku_server       | [2022-09-20 19:19:55,583] INFO None Identity: {"identity": {"account_number": "10001", "type": "User", "user": {"username": "user_dev", "email": "user_dev@foo.com", "is_org_admin": true}}, "entitlements": {"cost_management": {"is_entitled": true}, "sources": {"is_entitled": true}}}
koku_server       | [2022-09-20 19:19:55,741] INFO None enable_negative_filtering context: {'schema': 'acct1234567'}
koku_server       | [2022-09-20 19:19:55,742] INFO None     Negative Filtering Enabled acct1234567
koku_server       | [2022-09-20 19:19:55,784] INFO None query_exclusions: (OR: ('project_name__icontains', 'example-project-id'), ('project_id__icontains', 'example-project-id'))
koku_server       | [2022-09-20 19:19:55,867] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/reports/gcp/costs/?group_by[gcp_project]=*&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&exclude[gcp_project]=example-project-id', 'status': 200, 'request_id': 'DEVELOPMENT', 'account': '10001', 'org_id': '1234567', 'username': 'user_dev', 'is_admin': True, 'response_time': '288.42 ms'}
koku_server       | [2022-09-20 19:19:55,870] INFO None "GET /api/cost-management/v1/reports/gcp/costs/?group_by[gcp_project]=*&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&exclude[gcp_project]=example-project-id HTTP/1.1" 200 2266
```

**Check the Costs:**
Endpoint Returned:
```
"total": {
            "infrastructure": {
                "raw": {
                    "value": 1300.1806339,
                    "units": "USD"
                },
```
Overall - example-project-id = expected_cost
`5745347.77975477 - 5744047.59912087 = 1300.18063`


#### Additional things to consider:
 **Multi Exclude, Example:** 
- http://localhost:8000/api/cost-management/v1/reports/gcp/costs/?group_by[region]=*&filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&exclude[or:region]=australia-southeast1-a&exclude[or:region]=us-west1-a

**Tag Excludes, example:**
-  http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[tag:app]=*&exclude[tag:app]=analytics

**Multiple Tag Excludes, example:**
- http://localhost:8000/api/cost-management/v1/reports/aws/costs/?group_by[tag:app]=*&exclude[tag:app]=analytics&exclude[or:tag:app]=cost
## Notes

...
